src/util/messages.ml

```diff
@@ -3,6 +3,173 @@ open Pretty
+type integer = Overflow | DivByZero
+
+type cast = TypeMismatch
+
+type warning =
+  | Behavior of behavior
+  | Integer of integer
+  | Race
+  | Cast of cast
+  | Unknown of string
+  | Debug of string
+  | Analyzer
+
+module Warning =
+struct
+  type t = warning
+
+  module Behavior =
+  struct
+    type t = behavior
+
+    let create (e: t): warning = Behavior e
+    let undefined e: warning = create @@ Undefined e
+    let implementation (): warning = create @@ Implementation
+    let machine (): warning = create @@ Machine
+
+    module Undefined =
+    struct
+      type t = undefined_behavior
+
+      let create (e: t): warning = undefined e
+      let array_out_of_bounds e: warning = create @@ ArrayOutOfBounds e
+      let nullpointer_dereference (): warning = create @@ NullPointerDereference
+      let use_after_free (): warning = create @@ UseAfterFree
+
+      module ArrayOutOfBounds =
+      struct
+        type t = array_oob
+
+        let create (e: t): warning = array_out_of_bounds e
+        let past_end (): warning = create PastEnd
+        let before_start (): warning = create BeforeStart
+        let unknown (): warning = create Unknown
+
+        let show (e: t): string =
+          match e with
+          | PastEnd -> "[PastEnd]  Index is past the end of the array."
+          | BeforeStart -> "[BeforeStart] Index is before start of the array."
+          | Unknown -> "[Unknown] Not enough information about index."
+      end
+
+      let show (e: t): string =
+        match e with
+        | ArrayOutOfBounds e -> Printf.sprintf "[Array out of bounds] %s" (ArrayOutOfBounds.show e)
+        | NullPointerDereference -> "[Null pointer dereference]"
+        | UseAfterFree -> "[Use After Free]"
+    end
+
+    let show (e: t): string =
+      match e with
+      | Undefined u -> Printf.sprintf "[Undefined] %s" (Undefined.show u)
+      | Implementation -> "[Implementation]"
+      | Machine -> "[Machine]"
+  end
+
+  module Integer =
+  struct
+    type t = integer
+
+    let create (e: t): warning = Integer e
+    let overflow (): warning = create Overflow
+    let div_by_zero (): warning = create DivByZero
+
+    let show (e: t): string =
+      match e with
+      | Overflow -> "[Overflow]"
+      | DivByZero -> "[DivByZero]"
+  end
+
+  module Cast =
+  struct
+    type t = cast
+
+    let create (e: t): warning = Cast e
+    let type_mismatch (): warning = create TypeMismatch
+
+    let show (e: t): string =
+      match e with
+      | TypeMismatch -> "[TypeMismatch]"
+  end
+
+  let to_string e =
+    match e with
+    | Behavior _ -> "behavior"
+    | Integer _ -> "integer"
+    | Race -> "race"
+    | Cast _ -> "cast"
+    | Unknown msg -> "unknown"
+    | Debug msg -> "debug"
+    | Analyzer -> "analyzer"
+
+  let should_warn e =
+    get_bool ("warn." ^ (to_string e))
+
+  let show e =
+    match e with
+    | Behavior behavior -> Printf.sprintf "[Behavior] %s" (Behavior.show behavior)
+    | Integer _ -> "[Integer]"
```

Comment:
Is there a reason why the parameter is not used for `Integer` and `Cast`?

Reply:
just not finished yet

Reply:
they're now used